metadata/src/main/java/org/apache/kafka/metadata/placement/DefaultDirProvider.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.metadata.placement;
+
+import org.apache.kafka.common.Uuid;
+
+/**
+ * Provide the default directory for new partitions in a given broker.
+ */
+@FunctionalInterface
+public interface DefaultDirProvider {
+    Uuid defaultDir(int brokerId);
```

Comment:
So for a broker in JBOD mode (i.e. it has multiple directories), `defaultDir` would return `UNASSIGNED`, right? And when we're using an older MetadataVersion, we'll always get back `MIGRATING`, right?

It would be good to document that in the JavaDoc.

Reply:
I'll add documentation. But I don't think this should ever return `MIGRATING`, we simply do not call this provider unless the MetadataVersion supports direcotry assignments. The other result this might have is the single dir for a broker that supports JBOD but is registered with a single directory. Essentially this is so we can preset directory assignment to the only registered directory in a broker for new partitions or reassignments.

Reply:
> I'll add documentation. But I don't think this should ever return MIGRATING, we simply do not call this provider unless the MetadataVersion supports direcotry assignments.

Hmm... wouldn't it be simpler to handle the old metadata version case in the dir provider? If you handle it in the caller, that means you need to add more parameters to the `PartitionChangeBuilder`. That seems more complex.

Is there an advantage to the "handle MV externally" approach that I'm missing?

Reply:
The caller needs to handle it regarldless. If there's anything other than an empty list set in `directories`, serializing the record in a MV that does not yet support dir assignment triggers a `UnsupportedVersionException` with "Attempted to write a non-default directories at version ...".

I agree, I would rather not have to configure `PartitionChangeBuilder` with a `DefaultDirProvider`. But maybe I'm not quite picturing what you are suggesting.

Reply:
> The caller needs to handle it regarldless. If there's anything other than an empty list set in directories, serializing the record in a MV that does not yet support dir assignment triggers a UnsupportedVersionException with "Attempted to write a non-default directories at version ...".

In what I was proposing, the caller doesn't need to handle that, though, since it's not supposed to happen.

> I agree, I would rather not have to configure PartitionChangeBuilder with a DefaultDirProvider. But maybe I'm not quite picturing what you are suggesting.

Well, I was suggesting that the MV stuff be handled in the DefaultDirProvider rather than in PartitionChangeBuilder. But I guess the latter is already handling MV, so there isn't as much to be gained there. Fair enough.